contracts/TokenFactory.sol

```diff
@@ -173,12 +192,9 @@ contract TokenFactory is ReentrancyGuard, Ownable {
         uint256 _competitionId = competitionIds[tokenAddress];
         uint256 tokenCollateral = collateralById[_competitionId][tokenAddress];
 
-        uint256 remainingEthNeeded = FUNDING_GOAL - tokenCollateral;
         uint256 contributionWithoutFee = (valueToBuy * FEE_DENOMINATOR) /
             (FEE_DENOMINATOR + feePercent);
-        if (contributionWithoutFee > remainingEthNeeded) {
-            contributionWithoutFee = remainingEthNeeded;
-        }
+
```

Comment:
There are several issues with the code below in this function.

1. Let `r%` be fee percentage, and `p` be the amount the user paid. Currently fee is calculated as `p / (1+r) * r` where, and `contributionWithoutFee` (the collateral increment, i.e. ∆b)  is calculated as `p / (1+r)`. The two sums to `p`, since `∆b + fee = p / (1+r) + p / (1+r) * r = p`. However, the fee charged is not really `r%`. In reality it is 1 - 1 / (1+r) = r / (1+r), which is less than `r%`.
2.  Rather than pass in `valueToBuy`, the parameter passed in should be `amountPaid` (which is consistent with how it is currently used). The calculation of `contributionWithoutFee` should then be `amountPaid - fee`, where  `fee` should be simply  to be `calculateFee(amountPaid, feePercent)`
3. We should remove valueToReturn if it is not used
4. The public function `buy(address tokenAddress)` should also return the amount minted (so the client can accurately tell the user ahead of time via simulation-write)
5. There is no need to check `availableSupply` since we do not cap supply